### DISCUSSION TITLE
multiple validation steps

### DISCUSSION THREAD

**Discussion by yuvalsCG** [Lightning-AI/pytorch-lightning#11456] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I have a rather challenging use case. 
We are training 3d parametric models of objects. It is very beneficial for us to render the objects models during training to evaluate the models quality. 
Rendering is time-expensive action. We also like to get many relevant images from the `dataloader` to render the models on top of them.
It would be convenient for us if there would be another loop, in addition to `training_loop` and `validation_loop`,  let's call it `visualization_loop`, in which we can do visualizations.
The requirement from this visualization phase are:
1. We can control the rate it is being called (`check_every_n_epoch`)
2.  It is separated from validation_step. It has its own dataloader. **_Its dataloader loads heavy objects, and we don't want to iterate over it on every regular validation loop_**. (validation is a fast process which we want to do often, visualization is not). That is why the simple solution of multiple val_dataloader is not enough - they are being called in the same loop, same interval.

Seemingly, we would want to be able to create multiple `validation_loop`s which are separated by their functionality, their dataloaders and their rate of occurrence. 

I imagine something like this:

```
class Lit(pl.LightningModule):
    def training_step(self, batch: dict, batch_idx):
        "do something"

    def validation_step1(self, batch: dict, batch_idx):
        "do something"

    def validation_step2(self, batch: dict, batch_idx):
        "do the visualization"

trainer.fit(lit, train_loader, validation_dataloaders=[loader1, loader2], check_val_every_n_epoch=[1,20])

```

I was not able to find a solution to my problem in the current abstraction. I would be happy to hear proposed solutions and thoughts.

**Comment by rohitgr7** [comment]
hey @yuvalsCG !

one way you can achieve this:
```py
def validation_step1(...):
    ...

def validation_step2(...):
    ...

def validation_step(self, batch,  batch_idx, dataloader_idx):
    if dataloader_idx == 0:
        self.validation_step1(batch, batch_idx)
    elif dataloader_idx == 1:
        if (self.current_epoch + 1) % 20 != 0:
            return
        self.validation_step2(batch, batch_idx:
```
this will simplify the use-case of different validation_step hooks, but for the second dataloader, it might add a time overhead by retrieving the batch from the dataloader2 for some of the epochs where they aren't even used.
